Dynamic reconfiguration of service watchers

jolynch opened the pull request:
@igor47 @jnb 
This PR should enable Nerve to dynamically add and remove registrations in response to a SIGHUP
    
Why?
The zero downtime Nerve restart method that relies on starting duplicate nerves and relying on Synapse to de-duplicate registrations works until you start getting to hundreds of services per box. Then for every new service you schedule you have to send hundreds of ZK push updates across the entire infrastructure just to have Synapse go "oh yup it's the same". For us at Yelp this manifests as lots of Zookeeper "churn" (causing lots of snapshots) that has caused some instability for us.

How?
This patch changes the main loop to listen for a SIGHUP, and when it receives one it diffs the new configuration against the in memory state, launching and reaping watchers as needed.

Amazingly all the specs still pass without me changing them, which makes me think that I've got some tests to write. This diff is full of some subtle ordering constraints (signals + threads + methods with side effects galore), so I'd like to get some eyes on it as I'm writing specs.

I also think I may have fixed a bug with shutdown where the main thread used to wait forever (.join without a timeout) if it couldn't reap a watcher... now we give watchers 10s to go away before we Thread.kill them.

jolynch commented:
@EvanKrall, @solarkennedy this may be relevant to your interests.

In particular this change should make paasta services appear in SmartStack a lot faster (since we don't have to wait 60s, should only take ~2s to get a new registration now).

solarkennedy commented:
Wow, this is pretty great. I agree this would significantly improve the "response time" of nerve, and eliminate our dual-nerve upstart stuff (right?)

While this PR pushes that complexity into the nerve daemon, I think it is the "right" place to put it and allows it to benefit everyone who uses smartstack. It also puts nerve in line with how much production-level daemons operate, which is nice. (the reload-on-HUP behavior aspect)

I can't really comment on the Ruby. All this thread management looks like messy stuff and I have no experience with such things. :(

igor47 commented:
i feel like i've already suggested quite a refactor. does it make sense so far? i think given that we're doing much more complicated configuration management now, it makes sense to keep it all together in a single class.

do you want to take a stab at a re-architecture, and then i'll do another review?

jolynch commented:
Sure thing, that makes sense to me, I'll have a refactor with some specs asap :-)

jolynch commented:
@igor47 I think that I've refactored as you suggested. I've done some minimal testing locally and I'm going to start writing specs now ...

jolynch commented:
Hrm, I also need to think about how this is going to interact with process supervisors. Basically if "restart" becomes "send SIGHUP" then we're not sure if the config is going to blow up nerve as it is right now. Working on this issue.

igor47 commented on lib/nerve/configuration.rb:
```diff
@@ -0,0 +1,80 @@
+require 'optparse'
+
+module Nerve
+  class ConfigurationManager
```
to be more ruby-ish, this file name should match the object. move this file to `nerve/configuration_manager.rb` or rename this class to `Nerve::Configuration`

jolynch replied:
Done

igor47 commented on lib/nerve.rb:
```diff
@@ -15,35 +15,94 @@ class Nerve
       log.debug 'nerve: checking for required inputs'
       %w{instance_id services}.each do |required|
-        raise ArgumentError, "you need to specify required argument #{required}" unless opts[required]
+        raise ArgumentError, "you need to specify required argument #{required}" unless config[required]
```
this means that the main loop could bail with an `ArgumentError`; seems like we should maybe keep running with the old config in this case, but (a) i'm not sure if that's the right choice and (b) it's not even certain if we're already running, since this might be the first time we're loading the config (first loop).

not sure this is actionable now, but maybe it warrants a `TODO`

jolynch replied:
Yea, I mention this above as a concern. I'm currently trying out an apache like "check config" action but I'm not sure if I like that more than just logging an error and using the old configuration...

jolynch replied:
Yea, I ended up implementing a config check option similar to apachectl's "configtest"

A restart now looks like:
```
~/pg/nerve(branch:dynamic_reconfiguration*) » bundle exec bin/nerve -c $(CONFIG) -k && kill -SIGHUP $(PID)
I, [2016-03-04T21:29:00.416790 #8134]  INFO -- Nerve::Nerve: nerve: starting up!
I, [2016-03-04T21:29:00.416888 #8134]  INFO -- Nerve::Nerve: nerve: starting run
I, [2016-03-04T21:29:00.416913 #8134]  INFO -- Nerve::Nerve: nerve: loading config
I, [2016-03-04T21:29:00.418033 #8134]  INFO -- Nerve::Nerve: nerve: detected new config for your_http_service
I, [2016-03-04T21:29:00.418087 #8134]  INFO -- Nerve::Nerve: nerve: detected new config for your_tcp_service
I, [2016-03-04T21:29:00.418118 #8134]  INFO -- Nerve::Nerve: nerve: launching new watchers: ["your_http_service", "your_tcp_service"]
I, [2016-03-04T21:29:00.522977 #8134]  INFO -- Nerve::Nerve: nerve: reaping old watchers: []
I, [2016-03-04T21:29:00.523039 #8134]  INFO -- Nerve::Nerve: nerve: configuration check succeeded, exiting
W, [2016-03-04T21:29:00.523064 #8134]  WARN -- Nerve::Nerve: nerve: reaping all watchers
I, [2016-03-04T21:29:00.523083 #8134]  INFO -- Nerve::Nerve: nerve: exiting
exiting nerve
```

This isn't super optimal, but I think it's good enough for this revision.

igor47 commented on lib/nerve.rb:
```diff
@@ -15,35 +15,94 @@ class Nerve
-
-      log.debug 'nerve: completed init'
+      @instance_id = config['instance_id']
+      @services = config['services']
```
the distinction between `@services` and `@watchers` is hazy for me. maybe we could call this `watchers_desired` vs `watchers_existing` ?

jolynch replied:
I switched it to ``@watchers`` and ``@watchers_desired``. Yea, this naming gets especially confusing when I'm working on Synapse and Nerve branches (since they're all watchers...) heh lol

igor47 commented on lib/nerve.rb:
```diff
@@ -15,35 +15,94 @@ class Nerve
+            services_to_launch, services_to_reap = [], []
+
+            # Determine which watchers have changed their configuration
+            (@services.keys | @watchers.keys).each do |name|
```
not a fan of the `|` here. i would do this more functionally:

```ruby
watchers_to_reap = @watchers_existing.select{ |name, _| !@watchers_desired.keys.include?(name) }
do_the_reaping

watchers_to_launch = @watchers_desired.select{|name, _| !@watchers_existing.keys.include?(name) }
do_the_launching

watchers_to_update = @watchers_existing.select{|name, _| @watchers_desired.keys.include(name) }
etc...
```

jolynch replied:
I dig it, changed to be functional style.

jolynch commented:
Alright, I think I've incorporated your suggestions. Time for specs.

ubiquitousthey commented:
Is there any way I can help get this to a point it can be merged.  I would be very valuable.  We quite a bit of code in a supervisor that maintains two nerve instances to enable us to add services without taking existing services offline.

igor47 commented on lib/nerve.rb:
```diff
@@ -15,35 +15,103 @@ class Nerve
+            services_to_reap = @watchers.select{ |name, _|
+              !@watchers_desired.has_key?(name)
+            }.keys()
+            log.info "nerve: reaping old watchers: #{services_to_reap}"
```
could probably avoid outputting that log line unless we actually are going to reap something.

```ruby
unless services_to_reap.empty?
  log.info "..."
  services_to_reap.each {}
end
```

jolynch replied:
Done

igor47 commented on lib/nerve.rb:
```diff
@@ -15,35 +15,103 @@ class Nerve
+            services_to_launch = @watchers_desired.select{ |name, _|
+              !@watchers.has_key?(name)
+            }.keys()
+            log.info "nerve: launching new watchers: #{services_to_launch}"
```
likewise

jolynch replied:
Done

igor47 commented on lib/nerve.rb:
```diff
@@ -15,35 +15,103 @@ class Nerve
+            # Detect and update existing service watchers
+            services_to_update = @watchers.select { |name, _|
+              @watchers_desired.has_key?(name)
+            }.each { |name, _|
```
i would just start a new loop that's separate from the `select` for readability.
```ruby
services_to_update = select { }
services_to_update.each do |name, _|
...
end
```

also, in ruby, generally you use `{...}` for in-line blocks, and `do...end` for multi-line blocks

jolynch replied:
Ok, I've also refactored it a bit to be more consistent with the above loops.

igor47 commented on lib/nerve.rb:
```diff
@@ -15,35 +15,103 @@ class Nerve
+                @watcher_versions[temp_name] = @watcher_versions.delete(name)
+                log.info "nerve: launching new watcher for #{name}"
+                launch_watcher(name, @watchers_desired[name])
+                log.info "nerve: reaping old watcher #{temp_name}"
```
so, during this period, we're going to have the same service registered twice? is that... okay?

jolynch replied:
It should be ok because Synapse deduplicates registrations by host/port/name assuming that the Synapse service_watcher is properly calling [`set_backends`](https://github.com/airbnb/synapse/blob/master/lib/synapse/service_watcher/base.rb#L117-L121) which dedupes them for us.

But yea, I believe this is the only way to do a hitless reload (where the service doesn't de-register while nerve is changing).

igor47 commented on lib/nerve.rb:
```diff
@@ -15,35 +15,103 @@ class Nerve
+                log.info "nerve: launching new watcher for #{name}"
+                launch_watcher(name, @watchers_desired[name])
+                log.info "nerve: reaping old watcher #{temp_name}"
+                reap_watcher(temp_name) rescue "nerve: could not cleanly reap #{name}"
```
feels like the reaping semantics depend on the reporter used by the watcher. what if the new watcher's reporter creates an entry in a data store, assuming it owns it, and then the old watcher's reporter deletes that same entry? now we're in a seemingly bad place...

jolynch replied:
So I've thought about this a lot, and I think it's an acceptable risk because reporters by their nature _must_ be idempotent. There is no guarantee that we're not running two nerves with identical configurations at the same time (in fact we do this in production constantly). There is no guarantee two boxes are not running identical nerve configurations (which could run into the ownership issues you're talking about if the created state is not namespaced by host).

The zookeeper reporter does satisfy this requirement by using ephemeral nodes. The etcd reporter would I believe satisfy this through using create_in_order (but also the etcd reporter is sorta borked for other reasons anyways).

igor47 replied:
for example, if someone is using the leader election code in synapse, the new ephemeral node will still cause an election as it will be the highest sequence number.

it almost seems like if we want to be totally "correct" we should implement a hand-off protocol in the watchers -- maybe the watcher can be given an existing reporter?

i'm okay with shipping a "not-correct" version to get others unblocked, and then trying to figure out how to avoid the double-entry in a follow-up PR.

jolynch replied:
I have mixed feelings on this. On the one hand I feel like master failovers should be the "common" case so the application ought be able to deal with them. On the other hand I get that with something like mysql that might not be the case. I'll respond to the hand-off idea (which is definitely a good idea) in the main thread.

jolynch commented:
@ubiquitousthey This isn't too far, I just have to write some proper specs and work on the feedback igor47 just gave me.

I'll try to finish this up this weekend :-)

jolynch commented:
@igor47 I believe I've refactored to make it cleaner, along the way while testing I noticed a bug where I didn't properly handle service config changes due to how ruby does hash copies (rly the only way to do a deep copy is to use marshal...?).

Current status of this PR:
Functionality: I think it mostly works. I can manually change the configuration file in various ways, send SIGHUP and see the proper registrations be removed, added, changed, etc ... I can use the check mode to ensure that the config is valid before making Nerve load it (this is _technically_ a race condition but apparently Apache has this race condition as well and it works out ok).

Tests: I've written a basic test of the config_manager, but I still need to figure out how to properly test the SIGHUP changes. I'm working on those tests now.

My current concerns:
The launch/reap in the "update" section is hella racey. There is no guarantee that the launching thread has time to register the service before the old watcher is reaped. I'm considering using a thread local variable to signal to the main thread that a watcher has actually started, which we can optionally wait for in `launch_watcher`. Also tests of all these shenanigans.

igor47 commented:
i guess if we're concerned about the race conditions with registration and want to make sure that the service doesn't flap during reload, then now might be a good time to implement a hand-off? that is more correct and probably just as complex to implement as some sort of thread synchronization.

jolynch commented:
I'm concerned about race conditions on the launch/reap cycle because we've been bitten by that in production (where the "new" nerve didn't have time to register all services before the "old" nerve went away). The first priority of Nerve has to be to keep services registered otherwise Synapse can ruin your day.

I guess the way I look at it is that this PR is strictly better than the status quo at preventing churn (where if you change any registration all registrations on the box have to bounce), and it introduces a lot of useful machinery for moving towards a truly hitless world.

I agree that some kind of reporter owned hand-off would be superior, but I would like to do it in a follow up PR if possible as this is already a fairly risky change. Presumably the refactor for handoff strategy would be something like exposing a `handoff` method from ServiceWatchers and instead of launch/reap we'd call handoff.

igor47 commented:
that seems reasonable, carry on!

ubiquitousthey commented:
Just a thought.  What if you setup watchers and waited until you got matching ZK nodes or a timeout?

jolynch commented:
@ubiquitousthey I think that's exactly the right direction, namely we need to wait until the first check_and_report finishes in the watcher. The slightly tricky part is that by default we don't want to wait until that happens (e.g. when Nerve starts up we want to fire up 100 threads that all connect to ZK asap and setup ephem nodes), we only want to wait when we're replacing an existing service_watcher.

This isn't that hard with thread local variables (a la :finish), I just have to do it heh.

ubiquitousthey commented:
In Python or Java, I would use an Event for this kind of thread synchronization.  A little googling did not reveal such a thing in Ruby.  I did find a small implementation at https://emptysqua.re/blog/an-event-synchronization-primitive-for-ruby/ based on the Python implementation using Ruby's Mutex.

jolynch commented:
@igor47 I've added a spec for the main nerve application that I think exercises at least some of the interesting edge cases. I was struggling to actually test the threading interactions with standard rspec so I added the rspec_wait gem which helped a little. I still can't do the normal "expect this got called with this" kind of spec tests because those don't seem to work across threads. If you know of a better way that doesn't involve mocking out all the threads please let me know (which sorta defeats the point of the test IMO since all the logic is asserting that the threads get started and reaped appropriately).

If you don't see anything crazy in this I'd like to potentially merge it and start deploying it at Yelp in dev/staging to get some real-world experience with it. If that goes well and we make it to prod I'll circle back and cut a new release.

@ubiquitousthey  So ruby does have [ConditionVariables](http://ruby-doc.org/stdlib-2.0.0/libdoc/thread/rdoc/ConditionVariable.html) but I find them super awkward because you have to have a shared mutex. I've just signaled back to the main thread with another thread local variable, which seems to work reasonably well.

jolynch commented:
Ugh, crazy threading tests are introducing flakes ... I'll figure that out and loop back.

igor47 commented on spec/lib/nerve_spec.rb:
```diff
@@ -0,0 +1,139 @@
+      allow_any_instance_of(Nerve::ServiceWatcher).to receive(:run) {
+        # ServiceWatchers just infinite loop
+        loop do
+          Thread.current[:has_reported] |= true
```
do you mean `||=` here as well?

jolynch replied:
Yup, changed.

igor47 commented on spec/lib/nerve_spec.rb:
```diff
@@ -0,0 +1,139 @@
+
+  describe 'full application run' do
+    before(:each) {
+      allow_any_instance_of(Nerve::ServiceWatcher).to receive(:run) {
```
in general, `allow_any_instance_of` is a smell. rspec recommends not using it if possible: https://relishapp.com/rspec/rspec-mocks/docs/working-with-legacy-code/any-instance

any reason you can't just instantiate a specific watcher? you could maybe define a `SleepingWatcher` in specs which just implements this function.

jolynch replied:
Ok, yea I tried using class_doubles but that didn't seem to work.

igor47 commented:
i'm surprised that subject is cached across examples. maybe something else was the problem there?

also, i would avoid actually creating threads during the spec run. one way to do that is like so:

```ruby
before do
  allow(Thread).to receive(:new).and_yield
end
```

if you're testing stuff like "do we properly join the threads we start", i would do something like:

```ruby
let(:thread_obj) { instance_double(Thread) }
before do
  allow(Thread).to receive(:new) do
    yield
    return thread_obj
  end
end

it 'properly joins the thread' do
  expect(thread_obj).to receive(:join)
  my_thing_that_starts_a_thread()
end
```

in general, to create more determinism, you should just run everything sequentially. remember, your goal is not to test whether `Thread`s work -- the proper place to test that is in Ruby's own test suite, and we can assume that as a given. in your own specs, you just want to make sure you invoke the correct methods.

igor47 commented on lib/nerve/utils.rb:
```diff
@@ -4,5 +4,14 @@ def safe_run(command)
+      nap_time = seconds
+      while nap_time > 0
+        yield if block_given?
+        sleep [nap_time, 1].min
```
magic number `1` should be made into a constant

```ruby
module Utils
  RESPONSIVE_SLEEP_INTERVAL = 1  # seconds
  def responsive_sleep
    sleep RESPONSIVE_SLEEP_INTERVAL
    nap_time -= RESPONSIVE_SLEEP_INTERVAL
  end
end

jolynch replied:
Ok if I make it an argument to the function that defaults to some number?

Also is that break in a block thing at all idiomatic? I felt really dirty writing it... but it leads to a fairly concise calling convention.

igor47 replied:
oh. i didn't think about it. yeah, so, probably a better way would be to declare that the block you pass to `responsive_sleep` returns a value which determines if the sleep should terminate. e.g.:

```ruby
def responsive_sleep(seconds, &should_we_exit)
  while true
    sleep 1
    break if (should_we_exit && should_we_exit.call)
  end
end
```

jolynch replied:
dig it

jolynch commented:
Yea, I'm conflicted about this.

To give you some context, I originally started with mocking out the Threads (this is why I took so long writing specs, I wrote them 3 times) but then I realized that I wasn't testing much of the actual code, especially the coordination over thread local variables.

If I wanted confidence that the components like responsive sleep, thread local variables were working (e.g. writable by the main thread) etc ... the only way I build that confidence is by launching bonafide threads. The tolerances are sufficiently large (30s vs ~1s responsive sleeps) that I believe the likelihood of flakes outweigh the benefit from having "real" tests.

For what it's worth, we've found [real integration tests](https://github.com/Yelp/nerve-tools/blob/master/dockerfiles/itest/itest/itest.py) to be much more useful for testing nerve and synapse than traditional "pure" unit tests. That's why I ended up leaning towards more of an integration test than a unit test, because as much as philosophy says one thing, the point of the tests are to help prevent bugs and assert contracts about how the code works, and because Nerve is so intimately tied to Threading behaviour (which changes in different ruby versions fwiw), I sorta lean towards using real Threads in the overall test as well.

igor47 commented on lib/nerve.rb:
```diff
@@ -86,15 +165,47 @@ def run
+      return deep_copy.merge({'instance_id' => @instance_id, 'name' => name})
+    end
+
+    def launch_watcher(name, config, wait=false)
```
i would turn this final argument into an opt hash:

```ruby
def launch_watcher(name, config, opts = {})
  wait = opts[:wait] || false
end
```

then invocations are more clear; instead of `launch_watcher(name, config, true)` it would be `launch_watcher(name, config, :wait => true)`

jolynch replied:
dig it

igor47 commented on lib/nerve.rb:
```diff
@@ -86,15 +165,47 @@ def run
+      watcher = ServiceWatcher.new(watcher_config)
+      unless @config_manager.options[:check_config]
+        log.debug "nerve: launching service watcher #{name}"
+        watcher_thread = Thread.new { watcher.run }
```
maybe one way to simplify the testing is to move the thread management into the `ServiceWatcher` class? then, in the tests here, you simply assert that you call `watcher.run` or `watcher.stop`. you can test the thread-specific functionality using tests on the watcher -- including internal coordination.

you could then also avoid using thread-local variables, and instead use instance variables on the watcher instance. so, for instance, you would call `watcher.has_reported?` instead of checking `watcher_thread[:has_reported]`. internally, instead of `Thread.current[:has_reported]`, that method would just return `@has_reported`

jolynch replied:
That's an interesting idea. We'd have to have a Thread somewhere, so we'll have to deal with concurrent access to instance variables, but ... honestly that's not so bad (since we're just reading them, and they're both simple bools).

I'll poke at this and figure out how much of a refactor that would be.

igor47 commented on lib/nerve.rb:
```diff
@@ -86,15 +165,47 @@ def run
+      # Signal the watcher thread to exit
+      watcher_thread[:finish] = true
+
+      unclean_shutdown = watcher_thread.join(10).nil?
```
this functionality should really be delegated to the watcher itself. it should be sufficient to just call `watcher.stop` and know that it's going to work.

jolynch replied:
ok

igor47 commented on lib/nerve.rb:
```diff
@@ -86,15 +165,47 @@ def run
+      if unclean_shutdown
+        log.error "nerve: unclean shutdown of #{name}, killing thread"
+        Thread.kill(watcher_thread)
+        raise "Could not join #{watcher_thread}"
```
do we really want this exception? it's not caught, so will kill the entire process. it seems like the error is already handled when you call `Thread.kill`

jolynch replied:
Yea, pretty sure we just rescue it everywhere anyways... I'll change this.

igor47 commented:
completely agree with you on the integration tests -- i wrote such a suite too, in https://github.com/airbnb/smartstack-cookbook/blob/master/files/default/tests/minitest/test_test.rb, but i haven't maintained it at all :(

i disagree that you need to run actual threads to write unit tests for specific functions. the argument you made basically indicts the entire idea of unit tests (they're not "real" tests). for example, you list `responsive_sleep` as a component which requires actual threads to test, but i think this is a prime example of a component which can be unit tested with high confidence.

either way, the rspec suite is intended as a unit test suite, and i think it should be kept that way. if we want to integration-test the overall behavior, we should leave that to other test suites.

jolynch commented:
I definitely see where you're coming from, I'm not trying to argue against unit tests for small pieces of functionality. For example, I agree that responsive_sleep is imminently unit testable, and I can add a unit test for that if you want.

However, I do believe a test that uses real Threads is the more useful test of the Nerve Application than testing the interaction on mocks. The reality is that we _do_ care whether Ruby thread local variables are writable from the main thread in Ruby 2.2 vs 2.1 vs 1.9, and we do care about [backwards incompatible changes](https://github.com/airbnb/synapse/issues/137) to the Thread class. The whole purpose of the Nerve Application is to launch, reap, and signal with the ServiceWatcher threads, so let's make sure that actually works. I have isolated the Application from e.g. signal handling semantics and the realities of what ServiceWatcher threads actually do, but I feel I left enough reality in the suite to give me confidence when making changes to this behavior.

I guess what I'm trying to say is that I'm not arguing against unit tests, I'm claiming the unit test of the Nerve application is more useful if it uses real Threads. I agree that makes it more like an integration test than a pure unit test, but it's still testing an isolated piece of functionality and I don't believe that we'll see that many flakes as a result. Are you ok with us going with this and seeing what happens or do you strongly believe that for it to be a unit test it can't spawn threads?

jolynch commented:
Alright, I've got some train hacking to do. @igor47 I'll try to have the next revision ready for you to review tomorrow. Ignore my latest comment, if we push the threading down a level then the app doesn't have to care.

@ubiquitousthey I'm aiming to get this into a state it can be merged in tomorrow or maybe the next day, then it will take me a day to roll it out at Yelp, and if everything goes well I may be able to circle back by Friday to get a new version cut.

jolynch commented:
@igor47 I went ahead and did the refactor you suggested to push the threading into the ServiceWatcher. I believe that I've managed to do that without breaking compatibility with existing ServiceWatchers, and it made writing tests a bit easier as well.

The nested expect closed over an iteration counter is a tad strange, but I wasn't quite sure how else to test that subsequent iterations of the main loop handled the state changes right without it. I sorta like the resulting tests though.

👍 ?

bobtfish commented:
👍 from me.

igor47 commented on lib/nerve.rb:
```diff
@@ -86,15 +164,48 @@ def run
+      watcher = @watchers.delete(name)
+      @watcher_versions.delete(name)
+      shutdown_status = watcher.stop()
+      log.info "nerve: stopped #{name}"
```
wanna log the `shutdown_status` in this line? curious how often we're having unclean shutdowns

jolynch replied:
Sure. Yea I don't actually know if/why they're happening, but we have seen quite a few nerve processes just spin in sched_wait which I suspect may be related (that's is what we added the heartbeat code to detect). While refactoring I noticed this join without a timeout and that's definitely a possibility.

igor47 commented on lib/nerve.rb:
```diff
@@ -86,15 +164,48 @@ def run
+        @watchers[name] = watcher
+        if wait
+          log.info "nerve: waiting for watcher thread #{name} to report"
+          responsive_sleep(30) { !watcher.was_up.nil? || $EXIT }
```
magic numbers should be constants

jolynch replied:
Ok

igor47 commented on lib/nerve/service_watcher.rb:
```diff
@@ -57,32 +59,59 @@ def initialize(service={})
       @reporter.start()
 
-      until $EXIT
+      until $EXIT or @should_finish
```
maybe: 
```ruby
private
def watcher_should_exit
  $EXIT || @should_finish
end
```

then you can do `until watcher_should_exit` and also invoke that method (or just pass a reference to it via `method(:watcher_should_exit)`) in your `responsive_sleep`

fyi, you never want to use `or` or `and` in ruby -- stick with `||` and `&&`

jolynch replied:
Done. Yea I know about the or thing I just sometimes get my python and ruby confused lol. Cheers.

igor47 commented on spec/lib/nerve_spec.rb:
```diff
@@ -0,0 +1,186 @@
+require 'nerve'
+
+def make_mock_service_watcher
+    mock_service_watcher = double(Nerve::ServiceWatcher)
```
should this be an `instance_double` ? are we not using verifying doubles?

jolynch replied:
We are verifying doubles, e.g. if I say "allow(mock_service_watcher).to receive(:nope).and_return(true)" rspec gets mad at me with "expected no Exception, got #<RSpec::Mocks::MockExpectationError: Nerve::ServiceWatcher does not implement: nope>".

Apparently 'just' making it a double or instance_double doesn't allow it to receive expected method calls (this seems contrary to what the documentation says, but go ahead and try it).

igor47 commented on spec/lib/nerve_spec.rb:
```diff
@@ -0,0 +1,186 @@
+
+describe Nerve::Nerve do
+  let(:config_manager) { Nerve::ConfigurationManager.new() }
+  let(:mock_config_manager) { double() }
```
why do you need both a real and a fake config manager? should this be a real instance double, to make sure you're not invoking or stubbing methods that don't actually exist?

jolynch replied:
The real config manager is used during the test of the --check-config option (which validate that configuration is valid so that we can safely send a SIGHUP).

The fake config manager is used by the test of the regular run so that we are isolating the class under test more.

igor47 commented on spec/lib/nerve_spec.rb:
```diff
@@ -0,0 +1,186 @@
+      expect(nerve).to receive(:launch_watcher).once.with('service2', anything).and_call_original
+      expect(nerve).to receive(:reap_watcher).once.with('service2').and_call_original
+
+      expect(nerve).to receive(:heartbeat).exactly(iterations + 1).times {
```
for multi-line blocks, it's typical to use `do...end` instead of `{ }`

jolynch replied:
Done

igor47 commented:
the tests look *much* better now. thanks for doing the refactor! i left a few minor comments. my sense is the code is ready, although the tests could still be improved a bit (i'm particularly concerned about what seems to be a lack of verifying doubles, even though they seem to be turned on here: https://github.com/airbnb/nerve/blob/master/spec/spec_helper.rb#L18-L22).

jolynch commented:
@igor47 I feel I have integrated your suggestions and commented on the ones I didn't think needed to change. I believe this is pretty ready for merge. The ServiceWatcher refactor has made it clean and by switching to instance_doubles I checked that I did indeed get verified doubles. Also I've been doing fairly significant manual local testing. I kind of want to get this merged and deployed on real hardware.

Once it's deployed and I know it works on production I can loop back and cut a new release (if I find bugs I'll write regression tests on them and fix before cutting the release).

igor47 commented:
this PR is epic. nice work!

jolynch commented:
@igor47 is that a 👍 ?

jolynch commented:
Sweet, thanks @igor47 for all the review. I'm going to roll this out at Yelp and see what happens. If good things happen, I'll circle back and cut a release. I don't think that anything is non backwards compatible either on the watcher or config side.

jolynch commented:
@ubiquitousthey this should be available in 0.8.0, let me know if you run into any issues.

Also this should close out #62 ... only a year late

ubiquitousthey commented:
Thanks @jolynch.  I will give it a try.
